test/udruntime/include/updown.h

```diff
@@ -0,0 +1,1784 @@
+      if( it->second.free && size <= it->second.size ) {
+        // Check alignment
+        aligned_addr = it->first;
+        if( reinterpret_cast< uint64_t >( aligned_addr ) % 8 != 0 ) {
+          uint64_t alignment_offset =
+            8 - ( reinterpret_cast< uint64_t >( aligned_addr ) % 8 );
+          if( it->second.size >= size + alignment_offset ) {
+            aligned_addr =
+              static_cast< char* >( aligned_addr ) + alignment_offset;
+          } else {
+            continue;
+          }
+        }
+
+        UPDOWN_INFOMSG( "Found a region at 0x%lX",
+                        reinterpret_cast< uint64_t >( it->first ) );
+        used_reg = it;
+        break;
+      }
+    }
+    // Check if we found space
+    if( used_reg == regions_to_use->end() ) {
+      UPDOWN_ERROR( "Allocator run out of memory. Cannot allocate %lu bytes",
+                    size );
+      return nullptr;
+    }
+
+    // Calculate alignment offset
+    uint64_t alignment_offset = reinterpret_cast< uint64_t >( aligned_addr ) -
+                                reinterpret_cast< uint64_t >( used_reg->first );
+    void* alloc_addr = nullptr;
+
+    // needs to be aligned
+    if( alignment_offset != 0 ) {
+      ( *regions_to_use )[aligned_addr] = { size, false };
+      UPDOWN_INFOMSG( "Creating a new region 0x%lX with the size %lu for new "
+                      "aligned allocation",
+                      reinterpret_cast< uint64_t >( aligned_addr ),
+                      size );
+      alloc_addr = aligned_addr;
+      if( used_reg->second.size - alignment_offset > size ) {
+        // Create a new empty region for unused space
+        void* new_region_addr = static_cast< char* >( aligned_addr ) + size;
+        ( *regions_to_use )[new_region_addr] = {
+          used_reg->second.size - alignment_offset - size, true };
+        UPDOWN_INFOMSG(
+          "Creating a new region 0x%lX with the remaining size %lu",
+          reinterpret_cast< uint64_t >( new_region_addr ),
+          used_reg->second.size - alignment_offset - size );
+      }
+      // Create a new empty region for alignment
+      used_reg->second.size = alignment_offset;
+      used_reg->second.free = true;
+    } else {
+      // no alignment needed
+      // split the new region
+      uint64_t new_size     = used_reg->second.size - size;
+      used_reg->second.size = size;
+      used_reg->second.free = false;
+      alloc_addr            = used_reg->first;
+      // Create a new empty region
+      if( new_size != 0 ) {
+        void* new_region_addr = static_cast< char* >( used_reg->first ) + size;
+        ( *regions_to_use )[new_region_addr] = { new_size, true };
+        UPDOWN_INFOMSG(
+          "Creating a new region 0x%lX with the remaining size %lu",
+          reinterpret_cast< uint64_t >( new_region_addr ),
+          new_size );
+      }
+    }
+
+    UPDOWN_INFOMSG( "Returning region 0x%lX = {%lu, Used}",
```

Comment:
See above about `PRIu64`. `PRI` means `printf`, `u` means `unsigned` format, and `64` means 64-bit size. `"%" PRIu64` is the portable way to print `uint64_t`.